qiskit/aqua/operators/state_functions/circuit_state_fn.py

```diff
@@ -0,0 +1,348 @@
+     the probability of measuring
+    a given binary string b from a system with StateFn f is equal to the inner
+    product between f and m(b).
+
+    NOTE: State functions here are not restricted to wave functions, as there
+    is no requirement of normalization.
+    """
+
+    # TODO maybe break up into different classes for different fn definition primitives
+    # TODO allow normalization somehow?
+    def __init__(self,
+                 primitive: Union[QuantumCircuit, Instruction] = None,
+                 coeff: Union[int, float, complex, ParameterExpression] = 1.0,
+                 is_measurement: bool = False) -> None:
+        """
+            Args:
+                primitive: The operator primitive being wrapped.
+                coeff: A coefficient by which to multiply
+                 the state function.
+                is_measurement: Whether the StateFn is a measurement operator.
+
+            Raises:
+                TypeError: invalid parameters.
+        """
+        if isinstance(primitive, QuantumCircuit):
+            primitive = primitive.to_instruction()
+
+        if not isinstance(primitive, Instruction):
+            raise TypeError('CircuitStateFn can only be instantiated '
+                            'with Instruction, not {}'.format(type(primitive)))
+
+        super().__init__(primitive, coeff=coeff, is_measurement=is_measurement)
+
+    @staticmethod
+    def from_dict(density_dict: dict) -> OperatorBase:
+        """ from dictionary """
+        # If the dict is sparse (elements <= qubits), don't go
+        # building a statevector to pass to Qiskit's
+        # initializer, just create a sum.
+        if len(density_dict) <= len(list(density_dict.keys())[0]):
+            statefn_circuits = []
+            for bstr, prob in density_dict.items():
+                qc = QuantumCircuit(len(bstr))
+                # NOTE: Reversing endianness!!
+                for (index, bit) in enumerate(reversed(bstr)):
+                    if bit == '1':
+                        qc.x(index)
+                sf_circuit = CircuitStateFn(qc, coeff=prob)
+                statefn_circuits += [sf_circuit]
+            if len(statefn_circuits) == 1:
+                return statefn_circuits[0]
+            else:
+                return SummedOp(statefn_circuits)
+        else:
+            sf_dict = StateFn(density_dict)
+            return CircuitStateFn.from_vector(sf_dict.to_matrix())
+
+    @staticmethod
+    def from_vector(statevector: np.ndarray) -> OperatorBase:
+        """ from vector """
+        normalization_coeff = np.linalg.norm(statevector)
+        normalized_sv = statevector / normalization_coeff
+        if not np.all(np.abs(statevector) == statevector):
+            raise ValueError('Qiskit circuit Initializer cannot handle non-positive statevectors.')
+        return CircuitStateFn(Initialize(normalized_sv), coeff=normalization_coeff)
+
+    def get_primitives(self) -> set:
```

Comment:
Shouldn't the return typehint be `Set[str]`

Reply:
Fixed